hybridse/src/udf/dynamic_lib_manager.h

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2021 4Paradigm
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef HYBRIDSE_SRC_UDF_DYNAMIC_LIB_MANAGER_H_
+#define HYBRIDSE_SRC_UDF_DYNAMIC_LIB_MANAGER_H_
+
+#include <dlfcn.h>
+#include <string>
+#include <vector>
+#include <map>
+#include <mutex>
+#include "base/fe_status.h"
+
+namespace hybridse {
+namespace udf {
+
+struct DynamicLibHandle {
+    explicit DynamicLibHandle(void* ptr) {
+        handle = ptr;
+        ref_cnt = 1;
+    }
+    void* handle = nullptr;
+    uint32_t ref_cnt = 0;
+};
+
+class DynamicLibManager {
+ public:
+    DynamicLibManager() = default;
+
+    base::Status ExtractFunction(const std::string& name, bool is_aggregate, const std::string& file, std::vector<void*>* funs) {
+        if (funs == nullptr) {
+            return {common::kExternalUDFError, "funs is nullptr"};
+        }
+        std::shared_ptr<DynamicLibHandle> so_handle;
+        {
+            std::lock_guard<std::mutex> lock(mu_);
```

Comment:
should possible use read_write lock ?  like https://github.com/4paradigm/OpenMLDB/blob/e2b0b51bbcb7381c45723c96112ada171badaf22/src/statistics/query_response_time/deploy_query_response_time.h#L102-L103

Reply:
only used in create/drop function. it's not frequently to be called, so it's no matter what's mutex used